Sources/SwiftParser/Expressions.swift

```diff
@@ -997,18 +997,20 @@ extension Parser {
     // the token is an operator starts with '.', or the following token is '['.
     let rootType: RawTypeSyntax?
     if !self.at(prefix: ".") {
-      rootType = self.parseSimpleType(stopAtFirstPeriod: true)
+      rootType = self.parseSimpleType(allowMemberTypes: false)
     } else {
       rootType = nil
     }
 
     var components: [RawKeyPathComponentSyntax] = []
     var loopProgress = LoopProgressCondition()
     while self.hasProgressed(&loopProgress) {
-      // Check for a [] or .[] suffix. The latter is only permitted when there
-      // are no components.
+      // Check for a [] or .[] suffix. The latter is only permitted after the type name. Since we don't know what
+      // constitutes a nested type reference and what constitutes a type's member, we can only disallow it in the parser
+      // after seeing the first non-property component.
       if self.at(TokenSpec(.leftSquare, allowAtStartOfLine: false))
-        || (components.isEmpty && self.at(.period) && self.peek(isAt: .leftSquare))
+        || (components.allSatisfy({ $0.component.is(RawKeyPathPropertyComponentSyntax.self) }) && self.at(.period)
```

Comment:
I found `\T.?.?` is currently accepted, it also need this check. 

Also, could this `components.allSatisfy()` check a flag outside the loop? Like

Reply:
Updated to disallow these key paths. I found a fun example along the way that’s valid 😭. This prevents us from diagnosing the `.` as an unexpected token, which would generally be nicer diagnostics.

```swift
struct Foo {
  var bar: Int?
}

infix operator .?.

func .?.(_ x: AnyKeyPath, _ a: Int) {}

var blah = 2
\Foo?.?.bar.?.blah
```